core/app/c/[communitySlug]/pubs/PubSearchInput.tsx

```diff
@@ -0,0 +1,94 @@
+"use client";
+
+import React, { useDeferredValue, useEffect, useRef, useState } from "react";
+import { Search } from "lucide-react";
+import { parseAsInteger, parseAsString, useQueryStates } from "nuqs";
+import { useDebouncedCallback } from "use-debounce";
+
+import { KeyboardShortcutPriority, useKeyboardShortcut, usePlatformModifierKey } from "ui/hooks";
+import { Input } from "ui/input";
+import { cn } from "utils";
+
+type PubSearchProps = React.PropsWithChildren<{}>;
+
+const DEBOUNCE_TIME = 300;
+
+export const PubSearch = (props: PubSearchProps) => {
+	const [query, setQuery] = useQueryStates(
+		{
+			query: parseAsString.withDefault(""),
+			page: parseAsInteger.withDefault(1),
+		},
+		{
+			shallow: false,
+		}
+	);
+
+	// local input state for immediate UI responsiveness + sync with URL
+	// otherwise, when navigating back/forward or refreshing, the input will be empty
+	const [inputValue, setInputValue] = useState(query.query);
+
+	// deferred query to keep track of server updates
+	// without this, we can't only check if inputValue !== query.query,
+	// which only tells us that the debounce has happened
+	const deferredQuery = useDeferredValue(query.query);
```

Comment:
this is kind of mysterious to me, but this allows me to keep the body of the search at half-opacity until the content has fully refreshed. this could also be done by using `useTransition`.